zkevm-circuits/src/tx_circuit.rs

```diff
@@ -0,0 +1,542 @@
+//! The transaction circuit implementation.
+
+// Naming notes:
+// - *_be: Big-Endian bytes
+// - *_le: Little-Endian bytes
+
+mod sign_verify;
+
+use crate::util::Expr;
+use eth_types::{Address, Bytes, Field, ToBigEndian, ToLittleEndian, ToScalar, Word, U256};
+use ff::PrimeField;
+use group::GroupEncoding;
+use halo2_proofs::{
+    circuit::{AssignedCell, Layouter, Region, SimpleFloorPlanner},
+    plonk::{Advice, Circuit, Column, ConstraintSystem, Error},
+    poly::Rotation,
+};
+use itertools::Itertools;
+use lazy_static::lazy_static;
+use libsecp256k1;
+use log::error;
+use num::Integer;
+use num_bigint::BigUint;
+use rlp::RlpStream;
+use secp256k1::Secp256k1Affine;
+use sha3::{Digest, Keccak256};
+use sign_verify::{SignData, SignVerifyChip, SignVerifyConfig};
+pub use sign_verify::{POW_RAND_SIZE, VERIF_HEIGHT};
+use std::convert::TryInto;
+use std::marker::PhantomData;
+use subtle::CtOption;
+
+lazy_static! {
+    // Scalar
+    static ref SECP256K1_Q: BigUint = BigUint::from_slice(&[
+        0xd0364141, 0xbfd25e8c,
+        0xaf48a03b, 0xbaaedce6,
+        0xfffffffe, 0xffffffff,
+        0xffffffff, 0xffffffff,
+    ]);
+}
+
+/// Transaction to be verified by the TxCircuit
+#[derive(Clone, Default, Debug)]
+pub struct Transaction {
+    /// Sender address
+    pub from: Address,
+
+    /// Recipient address (None for contract creation)
+    pub to: Option<Address>,
+
+    /// Supplied gas
+    pub gas: U256,
+
+    /// Gas price
+    pub gas_price: U256,
+
+    /// Transfered value (None for no transfer)
+    pub value: U256,
+
+    /// The compiled code of a contract OR the first 4 bytes of the hash of the
+    /// invoked method signature and encoded parameters. For details see
+    /// Ethereum Contract ABI
+    pub data: Bytes,
+
+    /// Transaction nonce
+    pub nonce: U256,
+
+    /// "v" value of the transaction signature
+    pub v: u64,
+    /// "r" value of the transaction signature
+    pub r: U256,
+    /// "s" value of the transaction signature
+    pub s: U256,
+}
+
+fn random_linear_combine<F: Field>(bytes: [u8; 32], randomness: F) -> F {
+    crate::evm_circuit::util::Word::random_linear_combine(bytes, randomness)
+}
```

Comment:
So this is a very "stupid" function that allows me to write `random_linear_combine` instead of `Word::random_linear_combine`, which is shorter :sweat_smile: 

I'm not sure if this should belong to `gadgets` because it's not a gadget but an expression builder.  But I think it should be defined in `zkevm_circuits::util` instead of `zkevm_circuits::evm_circuit::util`